drivers/wifi/simplelink/simplelink_sockets.c

```diff
@@ -537,29 +550,37 @@ static int simplelink_poll(struct pollfd *fds, int nfds, int msecs)
 
 	for (i = 0; i < nfds; i++) {
 		fds[i].revents = 0;
-		fd = fds[i].fd;
-		if (fd < 0) {
+		if (fds[i].fd < 0) {
 			continue;
+		} else if ((fds[i].fd >= ARRAY_SIZE(simplelink_offload_fd)) ||
+			   (simplelink_offload_fd[fds[i].fd] == -1)) {
+			/* Non-offloaded socket, return an error. */
```

Comment:
This approach doesn't allow to mix up socket types, that's the limitation of offloading the entire `poll` call. 

For native sockets, we also have a less restrictive approach for `poll`, utilizing `ZFD_IOCTL_POLL_PREPARE` and `ZFD_IOCTL_POLL_UPDATE` `ioctl` requests. In this approach, every registered socket provider can provide a kernel object to observe with `ZFD_IOCTL_POLL_PREPARE` handler, and then update the poll results in `ZFD_IOCTL_POLL_UPDATE` handler. This allows to provide different socket types for a single `poll` call. 

If the offloaded socket implementation could easily provide such kernel objects, it's free to use this approach. For nRF/Simplelink though I found this difficult, simply because these engines offload entire `poll`/`select` calls.

Reply:
> This approach doesn't allow to mix up socket types, that's the limitation of offloading the entire poll call.

I'd formulate it a bit differently: this approach leaves handling of mixed socket types to the socket (offloading) handler. And the baseline of such handling would be of course to not allow **and** error out with appropriate error code. I just starting with a detailed review, but I hope we'll take that as a baseline indeed to check for mixed case and return error, instead of not handling it in explicit manner.

> If the offloaded socket implementation could easily provide such kernel objects, it's free to use this approach. For nRF/Simplelink though I found this difficult, simply because these engines offload entire poll/select calls.

For the purpose of discussion: the whole idea of poll() is to wait for *async* event on sockets. Hardware-wise, async events are implemented by interrupts. Which in turn translates to a callback at the lowest software level. So, the minimum requirement to support baseline-efficient handling of mixed socket types would be: an underlying offloading impl to provide means to register an "even callback". Just one callback, not per-socket. Of course filtering of event types is helpful. Then on Zephyr side of things, this callback would signal an event or semaphore object. And ZFD_IOCTL_POLL_PREPARE would return this single event/sema for every offloaded socket. And then, after getting an event, ZFD_IOCTL_POLL_UPDATE will take care of figuring out which socket(s) this event actually affected.

So, it shouldn't be that too hard to support mixed sockets if underlying provider supports the suitable low-level interface. Obviously, I don't for implementation of that any time soon. But it's helpful to discuss approaches to that, to validate our design.